Build changes for Spark 3.0.3, 3.0.4, 3.1.1, 3.1.2, 3.1.3, 3.1.1cdh and 3.0.1emr

tgravescs opened the pull request:
part of https://github.com/NVIDIA/spark-rapids/issues/3224

This is a lot of copied code. All the shim code for each of these was copied into the individual shims.  

Much of the core code from like SparkBaseShims was copied into the 3.0.x versions and for 3.1.1 I split that code out into its own SparkBaseShims with the changes from Spark311Shims and that got copied to the 3.1.x code, except for the cases it was overrident like ParquetFilters. The files should diff pretty well.

The other changes were to add build profiles.  For the shims/aggregator and shims, I created another profile called releaseOne that handles all the dependencies so we didn't need individual ones.
Unfortunately for cdh to build I had to make specific changes to various pom files to deal with dependencies there.

I also made the tests-spark310+ only run for Spark 3.1.1 right now because it directly references the shim version for 3.1.1  ParquetCachedBatchSerializer.  We want to move ParquetCachedBatchSerializer out of shims but can't do it until we switch the build system. For now it seems like just running it against 3.1.1 should be fine and fix when we move it around.

I've run all the unit tests for all these versions but have not run all the integration tests on all the final versions after upmerging, I'll kick some of those off.


<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->

tgravescs commented:
build

tgravescs commented:
ran mvn verify and had doc updates,

tgravescs commented:
build

tgravescs commented on pom.xml:
```diff
@@ -401,9 +499,13 @@
                   <artifactId>kryo-shaded</artifactId>
                 </exclusion>
+                <exclusion>
+                  <groupId>com.google.guava</groupId>
```
I need to make this conditional for just the new builds

tgravescs commented:
build

jlowe commented on sql-plugin/pom.xml:
```diff
@@ -91,11 +84,70 @@
+                <dependency>
+                    <groupId>org.apache.curator</groupId>
+                    <artifactId>curator-recipes</artifactId>
+                    <version>4.3.0.7.2.7.0-184</version>
```
By pulling in a different dependency version here, is there any concern with conflicts? Does it need to be shaded as we have done in other cases where we pull in a different dependency version? This comment applies to other places where the curator-recipes dependency was overridden as well.

tgravescs replied:
good question, I started just copying what they did in the shim layer so I'll check on this.

tgravescs replied:
I went ahead and changed all these to be provided as cdh environment should have these. 
 I also added a task for this in https://github.com/NVIDIA/spark-rapids/issues/3223 for us to check when we start combining these to make sure it doesn't leak/conflict.

tgravescs replied:
I also checked the final dist jar for cdh to make sure nothing else was in there and its good now

jlowe reviewed (commented):
It would be good to verify the contents of the distribution jar before/after this PR with the default build.

tgravescs commented on pom.xml:
```diff
@@ -216,36 +314,24 @@
-            </modules>
         </profile>
         <profile>
             <id>spark312tests</id>
```
no, but I didn't want to remove anything until we decided to switch over.

tgravescs commented on shims/spark312/src/main/scala/com/nvidia/spark/rapids/shims/spark312/ParquetCachedBatchSerializer.scala:
```diff
@@ -0,0 +1,1521 @@
+/*
```
so I agree, the plan is to move these out into someplace for spark 3.1+ but wasn't going to do that here, was going to do that separately.  I should probably file issues for the separate tasks.

tgravescs replied:
Note this also depends on if separate module or just separate source directory inside of sql-plugin and when we want to do that vs potentially breaking the old build.

tgravescs commented on sql-plugin/pom.xml:
```diff
@@ -44,18 +44,11 @@
           <groupId>org.apache.orc</groupId>
           <artifactId>orc-core</artifactId>
           <classifier>${orc.classifier}</classifier>
+          <version>1.5.8</version>
```
shouldn't even be needed, I think its left over from me debugging/figuring something out,

tgravescs commented on shims/spark301emr/src/main/scala/com/nvidia/spark/rapids/shims/spark301emr/GpuBroadcastNestedLoopJoinExec.scala:
```diff
@@ -0,0 +1,43 @@
+import org.apache.spark.sql.rapids.execution._
+
+/**
+ *  Spark 3.1 changed packages of BuildLeft, BuildRight, BuildSide
```
I'll clean them up

tgravescs commented on shims/spark301emr/src/main/scala/com/nvidia/spark/rapids/shims/spark301emr/GpuParquetScan.scala:
```diff
@@ -0,0 +1,72 @@
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+// FileScan changed in Spark 3.1.0 so need to compile in Shim
```
I'll clean them up

tgravescs commented on shims/spark301emr/src/main/scala/com/nvidia/spark/rapids/shims/spark301emr/GpuOrcScan.scala:
```diff
@@ -0,0 +1,70 @@
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.util.CaseInsensitiveStringMap
+
+// FileScan changed in Spark 3.1.0 so need to compile in Shim
```
I'll clean them up

tgravescs commented:
I think I addressed all the comments. I checked the final dist jar against one we push nightly and the only diffs I found were the extra shim files that we have now.

tgravescs commented:
build

jlowe commented on shims/spark311cdh/pom.xml:
```diff
@@ -96,6 +91,7 @@
             <groupId>org.apache.curator</groupId>
             <artifactId>curator-recipes</artifactId>
             <version>4.3.0.7.2.7.0-184</version>
+	        <scope>provided</scope>
```
Nit: Indentation is off here.

jlowe commented on udf-examples/pom.xml:
```diff
@@ -115,6 +115,40 @@
+                <groupId>org.apache.spark</groupId>
+                <artifactId>spark-sql_${scala.binary.version}</artifactId>
+                <version>${spark311cdh.version}</version>
+                <exclusions>
+                    <exclusion>
+                        <groupId>org.apache.curator</groupId>
+                        <artifactId>curator-recipes</artifactId>
+                    </exclusion>
+                </exclusions>
```
If we're pulling in the CDH version of the Spark jars, why would we need to override a dependency and replace it with something else?  Are the dependencies wrong on their jar?

tgravescs replied:
I assume so, this is what they had in the shim layer that they wrote.. I'll follow up with them to see if we can remove later

jlowe commented on udf-compiler/pom.xml:
```diff
@@ -141,6 +134,13 @@
+                    <groupId>com.google.guava</groupId>
+                    <artifactId>guava</artifactId>
+                    <version>28.0-jre</version>
+                    <scope>provided</scope>
```
Guava is notorious for being incompatible between the versions.  If the UDF compiled needs Guava we'll probably need to shade it to guarantee it gets the version it needs rather than hope the one it finds in the Spark env is compatible.

tgravescs replied:
changed to test scope and commented, asking Cloudera about it

tgravescs replied:
note it was just unit tests that needed it (thus the change to test scope - which is what I should have had it at in th efirst place)

tgravescs commented:
build
